### PR TITLE
Fixes for undefined names

### PR DESCRIPTION
**Pull request by PaulKlint**
This pull requests defines code fixes for undefined names. The beauty is that all users of TypePal can directly profit from this. I have implemented a Levenshtein distance metric (that closely follows the formal definition, but is still efficient), to compute the distance between a given variable in a set of possible IdRoles, with what is available in the given TModel.

My questions:
- @jurgenvinju : is this how you envisaged this stylistically?
- I now arbitrarily cutoff suggested names with an edit distance > 3. Should we make this a TypePal configuration parameter?
- Fixes are only generated for real errors, so the computational overhead is minimal.

Suggestions welcome (in particular about how test this in VsCode).

### CODE REVIEW

**Comment by DavyLandman** [comment]
Great stuff! I would like to have a feature to disable this functionality if desired via the config.

File: src/analysis/typepal/StringSimilarity.rsc [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+@synopsis{Tryadic minimum function on integers}
+int min(int a, int b, int c)
+=   a < b ? (a < c ? a : c) : (b < c ? b : c);
+
+@synopsis{Calculate the Levenshtein distance of 2 strings}
+int lev(str a, str b){
+    int sizea = size(a);
+    int sizeb = size(b);
+
+    @memo
```
Comment (DavyLandman) [inline src/analysis/typepal/StringSimilarity.rsc:16]: Should we configure the memo to timeout after a few minutes?

**Comment by jurgenvinju** [comment]
Yes this is exactly what I had in mind stylistically but I'd never dreamed this could be done generically for every typepal client so quickly. Wonder what else you have up your sleeve.

**Comment by jurgenvinju** [comment]
I agree with you both that an off-switch would be good for if we apply typepal in a non-interactive setting.

File: src/analysis/typepal/StringSimilarity.rsc (hunk 1 above)
Comment (PaulKlint) [reply src/analysis/typepal/StringSimilarity.rsc:16]: Yes will do

**Comment by jurgenvinju** [comment]
I've added an issue to fix the missing keyword parameter syntax issue. The only reason we don't have this is that we didn't want to change the syntax for a while. Internally all representations of function types already have keyword parameter names and types.

**Pull request merged**